NGAlert: Add integration tests for notification channels

codesome opened the pull request:
**What this PR does / why we need it**:

Adds integration test to test the following
1. POSTing a valid config for the channels is picked up properly
2. Routing is working fine (tested indireclty)
3. The payload is being generated and routed to right place for each channel.

Since email goes through a different path, I will add test for that in future PR and not this one. Similarly, telegram uses POST parameters, which is not trivial at the moment to add here, hence to be added in future PRs.

codesome committed:
NGAlert: Add integration tests for notification channels

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

gotjosh commented on pkg/tests/api/alerting/api_notification_channel_test.go:
```diff
@@ -0,0 +1,463 @@
+		fmt.Println(mockChannel.totalNotifications(), mockChannel.matchesExpNotifications(expNotifications))
+		return mockChannel.totalNotifications() == len(alertNames) &&
+			mockChannel.matchesExpNotifications(expNotifications)
+	}, 180*time.Second, 2*time.Second)
```
3 minutes seems a bit too much 😓  -- Is this the soonest we can get it to consistently? I would expect this to be a max of 20s.

codesome replied:
It was taking 50-60s. I will see what settings needs to be tweaked to speed it up.

codesome replied:
It was the `"group_wait": "0s"` missing in the config for the routes. It take around 20s now, so setting this to 25s.

gotjosh commented on pkg/tests/api/alerting/api_notification_channel_test.go:
```diff
@@ -0,0 +1,463 @@
+	cfg := apimodels.GettableUserConfig{}
+	require.NoError(t, json.Unmarshal(b, &cfg))
+
+	require.Equal(t, totalReceivers, len(cfg.AlertmanagerConfig.Receivers))
```
I'd like us to assert the actual payload here if possible, I'm certain there are a few things we'll need to fix as part of this PR once we do.

gotjosh reviewed (commented):
Looking 👌  - I'd like to see the payload of saved receivers before continuing with the review.

codesome commented:
PR is ready for review.

After a day of debugging on why dingding notifier was not working, I found out that the default template was not being picked up due to relative path issues. I have now changed how we store the default template in code - store it as Go variable instead of a template file.

I have added all remaining channels minus email and telegram since we still need to figure out stuff for them.

papagian commented on pkg/tests/api/alerting/api_notification_channel_test.go:
```diff
@@ -0,0 +1,956 @@
+func TestNotificationChannels(t *testing.T) {
+	dir, path := testinfra.CreateGrafDir(t, testinfra.GrafanaOpts{
+		EnableFeatureToggles: []string{"ngalert"},
+		AnonymousUserRole:    models.ROLE_EDITOR,
```
```suggestion
		DisableAnonymous: true,
```
In #33283 I have changed `createFolder()` (called below) and the folder is created with the main organisation instead of the anonymous organisation.
Therefore we have to disable the anonymous mode, [create a user](https://github.com/grafana/grafana/blob/b971d570fa17589388f2878cecffdf7a93a0d78f/pkg/tests/api/alerting/api_prometheus_test.go#L36) and [make authenticated requests](https://github.com/grafana/grafana/blob/b971d570fa17589388f2878cecffdf7a93a0d78f/pkg/tests/api/alerting/api_prometheus_test.go#L57).

codesome replied:
Perfect, thanks for this! I was wondering why it was giving 404 in CI and this should fix it.

papagian commented on pkg/tests/testinfra/testinfra.go:
```diff
@@ -202,6 +202,13 @@ func CreateGrafDir(t *testing.T, opts ...GrafanaOpts) (string, string) {
+	require.NoError(t, err)
+	_, err = alertingSect.NewKey("notification_timeout_seconds", "1")
+	require.NoError(t, err)
+	_, err = alertingSect.NewKey("evaluation_timeout_seconds", "1")
```
I don't think we use this configuration in `ngalert` evaluations yet. This should be part of grafana/alerting-squad#46

papagian reviewed (commented):
LGTM expect the `title_link` value in the test.

codesome commented on pkg/tests/api/alerting/api_notification_channel_test.go:
```diff
@@ -0,0 +1,961 @@
+		  "attachments": [
+			{
+			  "title": "Integration Test [FIRING:1]  (SlackAlert1)",
+			  "title_link": "TODO: rule URL",
```
Right, slack description also changed. Will update all.

codesome committed:
Fix the failing tests

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

gotjosh commented on pkg/services/ngalert/notifier/channels/utils.go:
```diff
@@ -32,3 +35,13 @@ func getTitleFromTemplateData(data *template.Data) string {
 	return title
 }
+
+// RemoveSystemLabels removes the labels added by Grafana from the alerts
```
I think it's best if we centralise this removal - In the end, I don't think those "system labels" are something we'd want and I want to avoid us having to scramble through multiple files to remove them.

How do you feel if we move it to the alert reception function? Can be part of a separate PR if you'd prefer (but I still think we need to fix the CI anyways)

codesome replied:
I moved it to `PutAlerts` now

gotjosh commented on pkg/tests/api/alerting/api_notification_channel_test.go:
```diff
@@ -0,0 +1,963 @@
+  "alertmanager_config": {
+    "route": {
+      "receiver": "slack_recv1",
+      "group_wait": "0s",
```
```suggestion
      "group_wait": "0s", // no group_wait time means the notification is sent as soon as it is received.
```

codesome replied:
Can't have comments in JSON :) so I will add this as part of the variable comment

gotjosh commented on pkg/tests/api/alerting/api_notification_channel_test.go:
```diff
@@ -0,0 +1,963 @@
+        "name": "email_recv",
+        "grafana_managed_receiver_configs": [
+          {
+            "id": 0,
+            "uid": "",
+            "name": "email_test",
+            "type": "email",
+            "isDefault": false,
+            "sendReminder": false,
+            "disableResolveMessage": false,
+            "frequency": "",
+            "created": "0001-01-01T00:00:00Z",
+            "updated": "0001-01-01T00:00:00Z",
+            "settings": {
+              "addresses": "test@email.com",
+              "singleEmail": true
+            },
+            "secureFields": {}
```
This is where I wanted to get on this PR, to me, this payload doesn't look right. I get the reason why it is this way though. In the configuration struct, the embedded field is a struct intended to be saved to the database - I think we should get away from this but I want to hear @papagian's thoughts on the matter and that we reach consensus here on what to do next.

However, I wouldn't expect the work to be done as part of this PR - An issue describing the problem/solution is more than acceptable.

codesome replied:
I am proposing removal of unwanted fields in this PR https://github.com/grafana/grafana/pull/34036

gotjosh commented on pkg/tests/api/alerting/api_notification_channel_test.go:
```diff
@@ -0,0 +1,963 @@
+var alertNames = []string{"DingDingAlert", "SlackAlert1", "SlackAlert2", "PagerdutyAlert", "TeamsAlert", "WebhookAlert"}
+
+func getRulesConfig(t *testing.T) string {
+	interval, err := model.ParseDuration("10s")
```
```suggestion
        t.Helper()
	interval, err := model.ParseDuration("10s")
```

nit: Can we move all this to the bottom? Test should be at the top of the file while helpers / fixtures at the bottom.

gotjosh commented on pkg/tests/api/alerting/api_notification_channel_test.go:
```diff
@@ -0,0 +1,963 @@
+
+	{
+		// There are no notification channel config initially.
+		alertsURL := fmt.Sprintf("http://grafana:password@%s/api/alertmanager/grafana/config/api/v1/alerts", grafanaListedAddr)
```
Is this right? I would expect the default configuration to be present...

codesome replied:
It passes, gives 404, so something wrong?

gotjosh replied:
This is meant to be, its my PR of #33539 that modifies this behaviour.

gotjosh commented on pkg/tests/api/alerting/api_notification_channel_test.go:
```diff
@@ -0,0 +1,963 @@
+}
+
+func (nc *mockNotificationChannel) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	nc.receivedNotificationsMtx.Lock()
```
```suggestion
        nc.t.Helper()
	nc.receivedNotificationsMtx.Lock()
```

gotjosh commented on pkg/tests/api/alerting/api_notification_channel_test.go:
```diff
@@ -0,0 +1,963 @@
+	key := fmt.Sprintf("%s/%s", urlParts[len(urlParts)-2], urlParts[len(urlParts)-1])
+	body := getBody(nc.t, req.Body)
+
+	fmt.Println("--- $$", key, body)
```
Is this supposed to be here?

codesome replied:
Nope, I added it to debug the CI failure. Will remove it once CI is green :)

gotjosh commented on pkg/tests/api/alerting/api_notification_channel_test.go:
```diff
@@ -0,0 +1,963 @@
+
+	fmt.Println("--- $$", key, body)
+
+	// Special handling for non-static data, like timestamp.
```
Sorry if this is a dumb question, but why do we have special handling in write path and then in the read path as well? I would expect us to handle special cases in one or the other (my preference is the read path as that way we preserve the original data)

codesome replied:
Moved it all to a single place - when checking expected to actual. And turns out I was missing 2 more special cases which I have fixed now.

gotjosh reviewed (approved):
Thank you for doing such a fantastic job,  most of my comments are strickly nits. I'm happy to take another look once the CI passes. 👏

codesome committed:
Fix review comments

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

codesome committed:
Override creation of rule UID, remove only namespace UID

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

codesome commented:
I have updated the PR to override rule UID creation, hence we need not remove it from the alert and can be available in template. I did not find a way for overriding namespace UID (fell short of time), so we are still removing it. But for now I think we can go ahead and look at overriding that later.

torkelo commented:
@codesome missing backport  (add backport v8.0.x for bot to automatically backport PR)
